### DISCUSSION TITLE
Azure AD authentication not working when access token verification is enabled.

### DISCUSSION THREAD

**Discussion by effusion** [quarkusio/quarkus#40658] [Q&A]
Hello,

we currently face the issue of not getting Azure Authentication working with `quarkus.oidc.authentication.verify-access-token` set to `true`.  

Our endpoint looks as follows:
```
@Path("/")
class IndexResource(val index: Template) {

    @GET
    @Produces(MediaType.TEXT_HTML)
    fun get(): TemplateInstance? {
        return index.instance()
    }
}
```

and the OIDC configuration like: 
```
quarkus.http.auth.permission.root.enabled=true
quarkus.http.auth.permission.root.paths=/*
quarkus.http.auth.permission.root.policy=authenticated
quarkus.oidc.auth-server-url=https://login.microsoftonline.com/<tenant-id>/v2.0
quarkus.oidc.application-type=hybrid
quarkus.oidc.token.refresh-expired=true
quarkus.oidc.authentication.redirect-path=/signin
quarkus.oidc.authentication.restore-path-after-redirect=true
quarkus.http.auth.permission.callback.paths=/signin
quarkus.http.auth.permission.callback.policy=authenticated
quarkus.oidc.authentication.scopes=profile,email
quarkus.oidc.authentication.cookie-same-site=lax
quarkus.oidc.roles.role-claim-path=roles
quarkus.oidc.token-state-manager.split-tokens=true
quarkus.oidc.authentication.verify-access-token=true
quarkus.oidc.token.customizer-name=azure-access-token-customizer
quarkus.oidc.token.issuer=any
quarkus.oidc.token.audience=any
```

The main problem at the moment is that even if `quarkus.oidc.token.audience` is set to `any`, we always get an exception that the `aud` claim does not match the expected value. For the property `quarkus.oidc.token.issuer` the value `any` works. Azure gives out tokens with different claims for id and access token.

Thanks to @sberyozkin help, I also found out about the https://quarkus.io/guides/security-oidc-bearer-token-authentication#jose4j-validator to customize claim validation. Unfortunately, this does not prevent the token verification from failing.

Did anyone else face this issues?

**Comment by sberyozkin** [comment]
Thanks @effusion, let me check... 

I'd like to clarify for everyone that what is described here, and which is what we briefly discussed with @effusion earlier, is the case where both Azure authorization code flow ID and access tokens have to be verified, with these two types of tokens having different issuer and audience claims.

The proposed plan has been to set both of these claims to `any`  but compensate with the extra verification of these claims in a custom validator, depending on a token type.

**Reply by sberyozkin** [reply]
@effusion Please see #40683, I have tests confirming that `audience=any` works but, more interestingly, you can set `audience` to contain a list of `,` separated values - which is simpler than using a custom `Jose4Validator` - but which you can still use as well if you want to make sure that a specific token type contains a specific audience value

**Reply by effusion** [reply]
@sberyozkin I tried it, but it still failed, rejecting the aud claim.

```
quarkus.oidc.authentication.verify-access-token=true
quarkus.oidc.token.customizer-name=azure-access-token-customizer
quarkus.oidc.token.audience=c81c4d29-cb5d-42f8-9a7d-dab0c020ac37,00000003-0000-0000-c000-000000000000
quarkus.oidc.token.issuer=any
```

It is always rejected with the message:
`Additional details: [[8] Audience (aud) claim [00000003-0000-0000-c000-000000000000] doesn't contain an acceptable identifier. Expected c81c4d29-cb5d-42f8-9a7d-dab0c020ac37 as an aud value.]`

Maybe the value itself causes a problem and not the configuration?

**Reply by sberyozkin** [reply]
@effusion I'm not sure what is going, may be you need to reload your application ? 
FYI, the test ID and access tokens I'm working with have the following `aud` format `{"aud":["somevalue"]}`, I've checked both of the tokens in your case, they have simple string like `aud` format `{"aud":"somevalue"}`.

So I've temp updated the test code to have `{"aud":"c81c4d29-cb5d-42f8-9a7d-dab0c020ac37"}` for the test ID token and `{"aud":"00000003-0000-0000-c000-000000000000"}` for the access token, and run them against the test with this expected configuration:

```
quarkus.oidc.token.audience=c81c4d29-cb5d-42f8-9a7d-dab0c020ac37,https://server.example.com
```

The ID token verification has passed but access token verification has failed, note the error message, here is a relevant fragment:

```
[[8] Audience (aud) claim [00000003-0000-0000-c000-000000000000] doesn't contain an acceptable identifier. Expected one of [c81c4d29-cb5d-42f8-9a7d-dab0c020ac37, https://server.example.com] as an aud value.]
```

Note the error message lists both configured values. In your case, despite configuring 2 audience values, the error message only includes the first one...

**Reply by effusion** [reply]
@sberyozkin Could you point me to the code where all the validation is happening? I can try and debug the code locally.

**Reply by sberyozkin** [reply]
@effusion It would be great, you can put a breakpoint at https://github.com/quarkusio/quarkus/blob/main/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcProvider.java#L226 and see what happens, it will be called twice, once for ID  token and once for the code flow access token

**Reply by effusion** [reply]
@sberyozkin, here are my findings from the debug session:
The variable always contains the value `c81c4d29-cb5d-42f8-9a7d-dab0c020ac37` as if the property `quarkus.oidc.token.audience` does not get adequately loaded.

In the Screenshot, you can see that `any` is for issuers, but the value `c81c4d29-cb5d-42f8-9a7d-dab0c020ac37` is for the audience.
The property was set to:
`quarkus.oidc.token.audience=any`

![oidc-debug](https://github.com/quarkusio/quarkus/assets/626784/d94a32a0-64fa-4084-89b6-5cff0ac06877)

If i can debug further, just let me know where.

**Reply by sberyozkin** [reply]
@effusion Thanks, it feels like it is some kind of deployment issue, the `c81c4d29-cb5d-42f8-9a7d-dab0c020ac37` you are seeing in the debug mode in the `OidcTenantConfig.Token` group can only be coming from `application.properties`, from `quarkus.oidc.token.audience`, so if you are still seeing `c81c4d29-cb5d-42f8-9a7d-dab0c020ac37`, despite resetting `quarkus.oidc.token.audience` to `any`, then it does look like you are seeing the stale configuration state.

Do you use env vars by any chance ?

**Reply by effusion** [reply]
We have an env file with a hardcoded audience value 😅. It works now with the list. I create the Josej4Validator to restrict the Issuer.

As an improvement, it would be nice if the issuer property is also a list like for audience.

**Reply by sberyozkin** [reply]
Hi @effusion , ok, glad you have got it working  
I have been thinking of introducing a dedicated confg group specifically for code flow access tokens to support such cases easier, typically multiple issues implies different oidc tenants should be involved. 

The fact both Azure tokens have different issuer and audience properties implies, that at least in theory, the AT is meant for the propagation 🙂.
I'll have a look later if listing multiple issuers may work
Thanks
